lib/mesh.h

```diff
@@ -0,0 +1,337 @@
+
+#include <array>
+#include <unordered_map>
+
+using RadiosityMesh = OpenMesh::PolyMesh_ArrayKernelT<>;
+
+// Property to store the vertices describing corners of a triangle
+// Since after subdivision, a triangle may consists of more than 3 vertices, we
+// have to keep track of the corners.
+using CornerVertices =
+    OpenMesh::FPropHandleT<std::array<RadiosityMesh::VertexHandle, 3>>;
+using CornerVerticesProperty =
+    OpenMesh::PropertyManager<CornerVertices, RadiosityMesh>;
+
+using VertexRadiosityHandle = OpenMesh::VPropHandleT<Color>;
+using VertexRadiosityHandleProperty =
+    OpenMesh::PropertyManager<VertexRadiosityHandle, RadiosityMesh>;
+
+using FaceRadiosityHandle = OpenMesh::FPropHandleT<Color>;
+using FaceRadiosityHandleProperty =
+    OpenMesh::PropertyManager<FaceRadiosityHandle, RadiosityMesh>;
+
+namespace detail {
+auto get_halfedge_handle(const RadiosityMesh& mesh,
+                         RadiosityMesh::FaceHandle face, size_t index) {
+    auto it = mesh.cfh_iter(face);
+    while (index--) {
+        it++;
+    }
+    return *it;
+}
+
+auto find_vertex(const RadiosityMesh& mesh, RadiosityMesh::FaceHandle face,
+                 const RadiosityMesh::Point& pt) {
+    size_t index = 0;
+    for (auto v : mesh.fv_range(face)) {
+        if (mesh.point(v) == pt) {
+            break;
+        }
+        index++;
+    }
+    return index;
+}
+} // namespace detail
+
+auto build_mesh(const Triangles& triangles) {
+    RadiosityMesh mesh;
+    // a vertex is uniquely determined by its coordinates and a normal
+    std::unordered_map<std::pair<Vec, Vec>, RadiosityMesh::VertexHandle> lookup;
+
+    auto corners_prop =
+        CornerVerticesProperty::createIfNotExists(mesh, "corner_vertices");
+
+    std::array<RadiosityMesh::VertexHandle, 3> vhandles;
+    for (const auto& tri : triangles) {
+        const Vec& a = tri.vertices[0];
+        const Vec& b = tri.vertices[1];
+        const Vec& c = tri.vertices[2];
+        const Vec& na = tri.normals[0];
+        const Vec& nb = tri.normals[1];
+        const Vec& nc = tri.normals[2];
+
+        auto ita = lookup.emplace(std::make_pair(a, na),
+                                  RadiosityMesh::VertexHandle());
+        if (ita.second) {
+            ita.first->second = mesh.add_vertex({a.x, a.y, a.z});
+        }
+        auto va = ita.first->second;
+
+        auto itb = lookup.emplace(std::make_pair(b, nb),
+                                  RadiosityMesh::VertexHandle());
+        if (itb.second) {
+            itb.first->second = mesh.add_vertex({b.x, b.y, b.z});
+        }
+        auto vb = itb.first->second;
+
+        auto itc = lookup.emplace(std::make_pair(c, nc),
+                                  RadiosityMesh::VertexHandle());
+        if (itc.second) {
+            itc.first->second = mesh.add_vertex({c.x, c.y, c.z});
+        }
+        auto vc = itc.first->second;
+
+        auto face = mesh.add_face(va, vb, vc);
+        corners_prop[face] = {va, vb, vc};
+    }
```

Comment:
Also, ich versuche einfach nur eine leere Vertex Handle in die Lookup-Table zu legen. Die mapped von den Paaren (position, normale) nach Vertex Handles. Die Rueckgabe ist ein paar; das zweite Argument ist ein bool, true, falls tatsächlich neuer Eintrag hinzugefügt wurde, sonst false. Haben wir einen neuen Eintrag, so kennen wir den Punkt noch nicht, und fügen den hinzu. 

1. So erstellen wir nur einmal einen Vertex für Dreieck mit gemeinsamen Ecken.
2. Das ist ein typischer Trick in C++. So spart man einen Lookup. Sonst müsste ja nach dem Vertex erst in der Tabelle suchen, und wenn der nicht da ist, einfügen. Das wären zwei Lookups.